lib/Sema/SemaExpr.cpp

```diff
@@ -8058,6 +8058,112 @@ Sema::CheckSingleAssignmentConstraints(QualType LHSType, ExprResult &CallerRHS,
   return result;
 }
 
+/// If a declaration has a Checked C bounds-safe interface
+/// attached to it, return the checked pointer type for the interface.
+/// Otherwise return a null QualType.
+///
+/// This falls into two cases:
+/// 1. The interface is a type annotation: return the type in the
+///    annotation.
+/// 2. The interface is a bounds expression.  This implies the checked
+/// type should be an _Array_ptr type.  Construct an _Array_ptr version
+/// of the unchecked ponter type for the declaration and return it.
+QualType Sema::GetInteropType(const ValueDecl *Decl) {
+  const DeclaratorDecl *TargetDecl = nullptr;
+  if (const FieldDecl *Field = dyn_cast<FieldDecl>(Decl))
+    TargetDecl = Field;
+  else if (const VarDecl *Var = dyn_cast<VarDecl>(Decl))
+    TargetDecl = Var;
+
+  if (TargetDecl) {
+    if (const BoundsExpr *Bounds = TargetDecl->getBoundsExpr()) {
```

Comment:
Yes, at some point, we'll have code that may need one or more bounds expressions and a bounds type annotation.   I agree that it makes sense to  prepare for that sooner rather than later.  Perhaps we should have separate accessors for the bounds expression and the bounds interface type annotation bounds type annotation, so that we can hid the concrete representation better?   This seems like something that shouldn't be part of this change, but should be done soon.